commit 9453bf03e838155e0eb31ce4b7e82dd35f79974c
Author: Susan Franklin <sfranklin>
Date:   Mon Feb 18 23:20:19 2008 +0000

    Bug 217272 - [prov] [ui] Need to prompt user when it is about installing a feature with install handlers

16	1	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
4	0	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIMessages.java
4	0	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
3	0	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKMessages.java
9	2	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
3	0	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
2	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
2	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
98	0	bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/UpdateManagerCompatibility.java
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
index 873dba5..ee877ec 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
@@ -11,8 +11,11 @@
 package org.eclipse.equinox.internal.p2.ui.admin;
 
 import org.eclipse.equinox.internal.provisional.p2.director.ProvisioningPlan;
-import org.eclipse.equinox.internal.provisional.p2.ui.*;
+import org.eclipse.equinox.internal.provisional.p2.ui.SimpleLicenseManager;
+import org.eclipse.equinox.internal.provisional.p2.ui.UpdateManagerCompatibility;
 import org.eclipse.equinox.internal.provisional.p2.ui.policy.*;
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.jface.dialogs.MessageDialog;
 import org.eclipse.jface.resource.ImageDescriptor;
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.ui.plugin.AbstractUIPlugin;
@@ -94,6 +97,18 @@ public class ProvAdminUIActivator extends AbstractUIPlugin {
 		if (planValidator == null)
 			planValidator = new IPlanValidator() {
 				public boolean continueWorkingWithPlan(ProvisioningPlan plan, Shell shell) {
+					if (plan == null)
+						return false;
+					// If the plan requires install handler support, we want to open the old update UI
+					if (UpdateManagerCompatibility.requiresInstallHandlerSupport(plan)) {
+						MessageDialog dialog = new MessageDialog(shell, ProvAdminUIMessages.ProvAdminUIActivator_UnsupportedInstallHandler, null, ProvAdminUIMessages.ProvAdminUIActivator_UnsupportedInstallHandlerMessage, MessageDialog.WARNING, new String[] {ProvAdminUIMessages.ProvAdminUIActivator_LaunchUpdateManager, ProvAdminUIMessages.ProvAdminUIActivator_ContinueAnyway, IDialogConstants.CANCEL_LABEL}, 0);
+						int ret = dialog.open();
+						if (ret == 1) // continue anyway
+							return true;
+						if (ret == 0)
+							UpdateManagerCompatibility.openInstaller();
+						return false;
+					}
 					return true;
 				}
 			};
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIMessages.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIMessages.java
index 8236690..7b30d5c 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIMessages.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIMessages.java
@@ -46,6 +46,10 @@ public class ProvAdminUIMessages extends NLS {
 	// Preferences
 	public static String InstalledIUPropertyPage_NoInfoAvailable;
 	public static String ProvAdminQueryProvider_UpdateQueryError;
+	public static String ProvAdminUIActivator_ContinueAnyway;
+	public static String ProvAdminUIActivator_LaunchUpdateManager;
+	public static String ProvAdminUIActivator_UnsupportedInstallHandler;
+	public static String ProvAdminUIActivator_UnsupportedInstallHandlerMessage;
 	public static String ProvisioningPrefPage_HideSystemRepos;
 	public static String ProvisioningPrefPage_ShowGroupsOnly;
 	public static String ProvisioningPrefPage_ShowInstallRootsOnly;
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
index 55a954e..4c913a6 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
@@ -10,6 +10,10 @@
 ###############################################################################
 
 ProvAdminQueryProvider_UpdateQueryError=Unexpected error while looking for updates
+ProvAdminUIActivator_ContinueAnyway=Continue anyway
+ProvAdminUIActivator_LaunchUpdateManager=Launch Update Manager...
+ProvAdminUIActivator_UnsupportedInstallHandler=Unsupported Install Handler
+ProvAdminUIActivator_UnsupportedInstallHandlerMessage=The IU you are trying to install represents a feature that defined an install handler.  This feature should only be installed with the old update UI.
 ProvisioningPrefPage_Description=Provisioning
 ProvisioningPrefPage_ShowGroupsOnly=Show only &group IU's in profile or repository views
 ProvisioningPrefPage_ShowInstallRootsOnly=Show only the install &roots in a profile
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKMessages.java b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKMessages.java
index f1a6b9b..7d9a42a 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKMessages.java
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKMessages.java
@@ -117,8 +117,11 @@ public class ProvSDKMessages extends NLS {
 	public static String ProvisioningPreferencePage_Prompt;
 	public static String ProvSDKQueryProvider_ErrorRetrievingProfile;
 	public static String ProvSDKUIActivator_ErrorWritingLicenseRegistry;
+	public static String ProvSDKUIActivator_LaunchUpdateManager;
 	public static String ProvSDKUIActivator_LicenseManagerReadError;
 	public static String ProvSDKUIActivator_NoSelfProfile;
 	public static String ProvSDKUIActivator_OpenWizardAnyway;
 	public static String ProvSDKUIActivator_Question;
+	public static String ProvSDKUIActivator_UnsupportedFeatureMessage;
+	public static String ProvSDKUIActivator_UnsupportedFeatureTitle;
 }
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
index ecdbc9a..d5637a9 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
@@ -23,8 +23,7 @@ import org.eclipse.equinox.internal.provisional.p2.ui.*;
 import org.eclipse.equinox.internal.provisional.p2.ui.operations.ProvisioningUtil;
 import org.eclipse.equinox.internal.provisional.p2.ui.policy.IPlanValidator;
 import org.eclipse.equinox.internal.provisional.p2.ui.policy.IQueryProvider;
-import org.eclipse.jface.dialogs.IDialogConstants;
-import org.eclipse.jface.dialogs.MessageDialogWithToggle;
+import org.eclipse.jface.dialogs.*;
 import org.eclipse.jface.resource.ImageDescriptor;
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.ui.plugin.AbstractUIPlugin;
@@ -199,8 +198,16 @@ public class ProvSDKUIActivator extends AbstractUIPlugin {
 				public boolean continueWorkingWithPlan(ProvisioningPlan plan, Shell shell) {
 					if (plan == null)
 						return false;
+					// If the plan requires install handler support, we want to open the old update UI
+					if (UpdateManagerCompatibility.requiresInstallHandlerSupport(plan)) {
+						MessageDialog dialog = new MessageDialog(shell, ProvSDKMessages.ProvSDKUIActivator_UnsupportedFeatureTitle, null, ProvSDKMessages.ProvSDKUIActivator_UnsupportedFeatureMessage, MessageDialog.WARNING, new String[] {ProvSDKMessages.ProvSDKUIActivator_LaunchUpdateManager, IDialogConstants.CANCEL_LABEL}, 0);
+						if (dialog.open() == 0)
+							UpdateManagerCompatibility.openInstaller();
+						return false;
+					}
 					if (plan.getStatus().isOK())
 						return true;
+
 					// Special case those statuses where we would never want to open a wizard
 					if (plan.getStatus().getCode() == IStatusCodes.NOTHING_TO_UPDATE) {
 						ProvUI.reportStatus(plan.getStatus(), StatusManager.BLOCK);
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
index a07856e..59c98f6 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
@@ -90,10 +90,13 @@ AutomaticUpdatesPreferencePage_RemindOnce=Notify me &once about updates
 AutomaticUpdatesJob_Updates=Updates
 ProvSDKQueryProvider_ErrorRetrievingProfile=Error while retrieving profile {0}
 ProvSDKUIActivator_ErrorWritingLicenseRegistry=Error writing license registry.  Accepted licenses will not be remembered.
+ProvSDKUIActivator_LaunchUpdateManager=Launch Update Manager...
 ProvSDKUIActivator_LicenseManagerReadError=Error reading license registry.  Accepted licenses will not be remembered.
 ProvSDKUIActivator_NoSelfProfile=Could not locate the running profile instance
 ProvSDKUIActivator_OpenWizardAnyway=The software items you selected may not be valid with your current installation.  Do you want open the wizard anyway to review the selections?
 ProvSDKUIActivator_Question=Question
+ProvSDKUIActivator_UnsupportedFeatureMessage=The feature that you are trying to install uses older Update Manager mechanisms.  This software can only be installed by the old update manager.
+ProvSDKUIActivator_UnsupportedFeatureTitle=Unsupported Feature
 ProvisioningPreferencePage_Always=&Open the wizard anyway to review or change selections
 ProvisioningPreferencePage_BrowsingPrefsGroup=Browsing for updates
 ProvisioningPreferencePage_ShowLatestVersions=Show only the &latest versions of available software
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
index ba8d864..f740787 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/ProvUIMessages.java
@@ -174,6 +174,8 @@ public class ProvUIMessages extends NLS {
 	public static String UpdateAndInstallSelectionDialog_DeselectAllLabel;
 	public static String UpdateAndInstallSelectionDialog_SelectAllLabel;
 	public static String UpdateAndInstallGroup_Refresh;
+	public static String UpdateManagerCompatibility_UnableToOpenFindAndInstall;
+	public static String UpdateManagerCompatibility_UnableToOpenManageConfiguration;
 	public static String UpdateOperation_NothingToUpdate;
 
 }
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
index 864dda4..295bc1d 100644
--- a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/p2/ui/messages.properties
@@ -115,6 +115,8 @@ UpdateDialog_AssemblingUpdatesProgress=Assembling updates
 UpdateAction_UpdatesAvailableTitle=Available Updates
 UpdateOperation_NothingToUpdate=There is nothing to update.
 UpdateAndInstallGroup_Refresh=&Refresh
+UpdateManagerCompatibility_UnableToOpenFindAndInstall=Unable to open the Update Manager Find and Install Wizard.
+UpdateManagerCompatibility_UnableToOpenManageConfiguration=Unable to open the Update Manager Configuration Manager.
 CategoryElementCollector_Uncategorized=Uncategorized
 ChooseProfileDialog_DefaultDialogTitle=Select a profile
 InstallDialog_InstallSelectionMessage=Review and confirm that the checked items will be installed.
diff --git a/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/UpdateManagerCompatibility.java b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/UpdateManagerCompatibility.java
new file mode 100644
index 0000000..f0e1bed
--- /dev/null
+++ b/bundles/org.eclipse.equinox.p2.ui/src/org/eclipse/equinox/internal/provisional/p2/ui/UpdateManagerCompatibility.java
@@ -0,0 +1,98 @@
+/*******************************************************************************
+ * Copyright (c) 2008 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.equinox.internal.provisional.p2.ui;
+
+import org.eclipse.core.commands.*;
+import org.eclipse.core.commands.common.NotDefinedException;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.equinox.internal.p2.ui.ProvUIActivator;
+import org.eclipse.equinox.internal.p2.ui.ProvUIMessages;
+import org.eclipse.equinox.internal.provisional.p2.director.ProvisioningPlan;
+import org.eclipse.equinox.internal.provisional.p2.engine.InstallableUnitOperand;
+import org.eclipse.equinox.internal.provisional.p2.engine.Operand;
+import org.eclipse.equinox.internal.provisional.p2.metadata.IInstallableUnit;
+import org.eclipse.ui.PlatformUI;
+import org.eclipse.ui.commands.ICommandService;
+import org.eclipse.ui.handlers.IHandlerService;
+import org.eclipse.ui.statushandlers.StatusManager;
+
+/**
+ * Utility methods involving compatibility with the Eclipse Update Manager.
+ * 
+ * @since 3.4
+ *
+ */
+public class UpdateManagerCompatibility {
+
+	// This value was copied from MetadataGeneratorHelper.  Must be the same.
+	private static final String ECLIPSE_INSTALL_HANDLER_PROP = "org.eclipse.update.installHandler"; //$NON-NLS-1$
+	// These values rely on the command markup in org.eclipse.ui.ide that defines the update commands
+	private static final String UPDATE_MANAGER_FIND_AND_INSTALL = "org.eclipse.ui.update.findAndInstallUpdates"; //$NON-NLS-1$
+	private static final String UPDATE_MANAGER_MANAGE_CONFIGURATION = "org.eclipse.ui.update.manageConfiguration"; //$NON-NLS-1$
+
+	public static boolean requiresInstallHandlerSupport(ProvisioningPlan plan) {
+		Operand[] operands = plan.getOperands();
+		for (int i = 0; i < operands.length; i++) {
+			if (operands[i] instanceof InstallableUnitOperand) {
+				IInstallableUnit iu = ((InstallableUnitOperand) operands[i]).second();
+				if (iu != null && iu.getProperty(ECLIPSE_INSTALL_HANDLER_PROP) != null)
+					return true;
+			}
+		}
+		return false;
+
+	}
+
+	/**
+	 * Open the old UpdateManager installer UI using the specified shell. 
+	 * We do not call the UpdateManagerUI class directly because we want to be able to be configured 
+	 * without requiring those plug-ins.  Instead, we invoke a known command.
+	 */
+	public static void openInstaller() {
+		runCommand(UPDATE_MANAGER_FIND_AND_INSTALL, ProvUIMessages.UpdateManagerCompatibility_UnableToOpenFindAndInstall);
+	}
+
+	/**
+	 * Open the old UpdateManager configuration manager UI using the specified shell. 
+	 * We do not call the UpdateManagerUI class directly because we want to be able to be configured 
+	 * without requiring those plug-ins.  Instead, we invoke a known command.
+	 */
+	public static void openConfigurationManager() {
+		runCommand(UPDATE_MANAGER_MANAGE_CONFIGURATION, ProvUIMessages.UpdateManagerCompatibility_UnableToOpenManageConfiguration);
+	}
+
+	private static void runCommand(String commandId, String errorMessage) {
+		ICommandService commandService = (ICommandService) PlatformUI.getWorkbench().getService(ICommandService.class);
+		Command command = commandService.getCommand(commandId);
+		if (!command.isDefined()) {
+			return;
+		}
+		IHandlerService handlerService = (IHandlerService) PlatformUI.getWorkbench().getService(IHandlerService.class);
+		try {
+			handlerService.executeCommand(commandId, null);
+		} catch (ExecutionException e) {
+			reportFail(errorMessage, e);
+		} catch (NotDefinedException e) {
+			reportFail(errorMessage, e);
+		} catch (NotEnabledException e) {
+			reportFail(errorMessage, e);
+		} catch (NotHandledException e) {
+			reportFail(errorMessage, e);
+		}
+	}
+
+	private static void reportFail(String message, Throwable t) {
+		Status failStatus = new Status(IStatus.ERROR, ProvUIActivator.PLUGIN_ID, message, t);
+		ProvUI.reportStatus(failStatus, StatusManager.BLOCK | StatusManager.LOG);
+
+	}
+}
\ No newline at end of file
commit a1368410e1eecb9b3d237c0da64abf63f033a796
Author: Susan Franklin <sfranklin>
Date:   Thu Feb 21 18:05:27 2008 +0000

    Bug 217272 - [prov] [ui] Need to prompt user when it is about installing a feature with install handlers

6	1	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
3	3	bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
6	1	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
5	3	bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
index ee877ec..8b724e5 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/ProvAdminUIActivator.java
@@ -17,6 +17,7 @@ import org.eclipse.equinox.internal.provisional.p2.ui.policy.*;
 import org.eclipse.jface.dialogs.IDialogConstants;
 import org.eclipse.jface.dialogs.MessageDialog;
 import org.eclipse.jface.resource.ImageDescriptor;
+import org.eclipse.swt.custom.BusyIndicator;
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.ui.plugin.AbstractUIPlugin;
 import org.osgi.framework.BundleContext;
@@ -106,7 +107,11 @@ public class ProvAdminUIActivator extends AbstractUIPlugin {
 						if (ret == 1) // continue anyway
 							return true;
 						if (ret == 0)
-							UpdateManagerCompatibility.openInstaller();
+							BusyIndicator.showWhile(shell.getDisplay(), new Runnable() {
+								public void run() {
+									UpdateManagerCompatibility.openInstaller();
+								}
+							});
 						return false;
 					}
 					return true;
diff --git a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
index 4c913a6..6790c6c 100644
--- a/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui.admin/src/org/eclipse/equinox/internal/p2/ui/admin/messages.properties
@@ -10,10 +10,10 @@
 ###############################################################################
 
 ProvAdminQueryProvider_UpdateQueryError=Unexpected error while looking for updates
-ProvAdminUIActivator_ContinueAnyway=Continue anyway
-ProvAdminUIActivator_LaunchUpdateManager=Launch Update Manager...
+ProvAdminUIActivator_ContinueAnyway=Continue
+ProvAdminUIActivator_LaunchUpdateManager=Launch UM...
 ProvAdminUIActivator_UnsupportedInstallHandler=Unsupported Install Handler
-ProvAdminUIActivator_UnsupportedInstallHandlerMessage=The IU you are trying to install represents a feature that defined an install handler.  This feature should only be installed with the old update UI.
+ProvAdminUIActivator_UnsupportedInstallHandlerMessage=The feature you are trying to install has defined an Update Manager install handler.  You may launch the old Update Manager in order to get the install handler support, or continue the install and ignore the install handler.
 ProvisioningPrefPage_Description=Provisioning
 ProvisioningPrefPage_ShowGroupsOnly=Show only &group IU's in profile or repository views
 ProvisioningPrefPage_ShowInstallRootsOnly=Show only the install &roots in a profile
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
index d5637a9..b250d45 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/ProvSDKUIActivator.java
@@ -25,6 +25,7 @@ import org.eclipse.equinox.internal.provisional.p2.ui.policy.IPlanValidator;
 import org.eclipse.equinox.internal.provisional.p2.ui.policy.IQueryProvider;
 import org.eclipse.jface.dialogs.*;
 import org.eclipse.jface.resource.ImageDescriptor;
+import org.eclipse.swt.custom.BusyIndicator;
 import org.eclipse.swt.widgets.Shell;
 import org.eclipse.ui.plugin.AbstractUIPlugin;
 import org.eclipse.ui.statushandlers.StatusManager;
@@ -202,7 +203,11 @@ public class ProvSDKUIActivator extends AbstractUIPlugin {
 					if (UpdateManagerCompatibility.requiresInstallHandlerSupport(plan)) {
 						MessageDialog dialog = new MessageDialog(shell, ProvSDKMessages.ProvSDKUIActivator_UnsupportedFeatureTitle, null, ProvSDKMessages.ProvSDKUIActivator_UnsupportedFeatureMessage, MessageDialog.WARNING, new String[] {ProvSDKMessages.ProvSDKUIActivator_LaunchUpdateManager, IDialogConstants.CANCEL_LABEL}, 0);
 						if (dialog.open() == 0)
-							UpdateManagerCompatibility.openInstaller();
+							BusyIndicator.showWhile(shell.getDisplay(), new Runnable() {
+								public void run() {
+									UpdateManagerCompatibility.openInstaller();
+								}
+							});
 						return false;
 					}
 					if (plan.getStatus().isOK())
diff --git a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
index 59c98f6..fe1c939 100644
--- a/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
+++ b/bundles/org.eclipse.equinox.p2.ui.sdk/src/org/eclipse/equinox/internal/p2/ui/sdk/messages.properties
@@ -90,13 +90,15 @@ AutomaticUpdatesPreferencePage_RemindOnce=Notify me &once about updates
 AutomaticUpdatesJob_Updates=Updates
 ProvSDKQueryProvider_ErrorRetrievingProfile=Error while retrieving profile {0}
 ProvSDKUIActivator_ErrorWritingLicenseRegistry=Error writing license registry.  Accepted licenses will not be remembered.
-ProvSDKUIActivator_LaunchUpdateManager=Launch Update Manager...
+ProvSDKUIActivator_LaunchUpdateManager=Launch...
 ProvSDKUIActivator_LicenseManagerReadError=Error reading license registry.  Accepted licenses will not be remembered.
 ProvSDKUIActivator_NoSelfProfile=Could not locate the running profile instance
 ProvSDKUIActivator_OpenWizardAnyway=The software items you selected may not be valid with your current installation.  Do you want open the wizard anyway to review the selections?
 ProvSDKUIActivator_Question=Question
-ProvSDKUIActivator_UnsupportedFeatureMessage=The feature that you are trying to install uses older Update Manager mechanisms.  This software can only be installed by the old update manager.
-ProvSDKUIActivator_UnsupportedFeatureTitle=Unsupported Feature
+ProvSDKUIActivator_UnsupportedFeatureMessage=A feature that you have selected uses install procedures that are not compatible with the current installation support.  This feature can only be installed by the older Update Manager.  \
+\
+Do you want to launch the older Update Manager?
+ProvSDKUIActivator_UnsupportedFeatureTitle=Unsupported Install
 ProvisioningPreferencePage_Always=&Open the wizard anyway to review or change selections
 ProvisioningPreferencePage_BrowsingPrefsGroup=Browsing for updates
 ProvisioningPreferencePage_ShowLatestVersions=Show only the &latest versions of available software
